### PR TITLE
AFL capability added to common.sh, corresponding changes to build.sh for LCMS and re2

### PR DESCRIPTION
**Pull request by dnoursi**
To build with afl, call:

`./build.sh afl`

Otherwise, both of these are equivalent:

```
./build.sh libfuzzer
./build.sh
```

We now name binaries "(name)_afl" in AFL mode, but maintain the status quo for libfuzzer; this way backward compatibility is maintained.

### CODE REVIEW

**Comment by dnoursi** [comment]
Pushed some new code regarding "target_files()", @morehouse

**Comment by morehouse** [comment]
I like this better than the previous version -- seems cleaner to call functions defined in common.sh than to require common.sh to be included at a specific point in the code.  Interested in what @kcc thinks of this.

File: common.sh [hunk 1]
```diff
@@ -32,6 +47,24 @@ get_svn_revision() {
   [ ! -e $TO_DIR ] && svn co -r$SVN_REVISION $SVN_REPO $TO_DIR
 }
 
+build_afl() {
+  $CC -c -w $AFL_SRC/llvm_mode/afl-llvm-rt.o.c
+  $CXX -g -fsanitize-coverage=trace-pc-guard $TARGET_FILE $TARGET_INCLUDE -c
+
+  UNIQUE_BUILD="$AFL_DRIVER afl-llvm-rt.o.o ${TARGET_NAME}.o $UNIQUE_BUILD"
```
Comment (kcc) [inline common.sh]: again, you are using TARGET_NAME defined in build.sh. Please don't. 
More generally, avoid circular deps between .sh files -- it gets too messy too soon. 
both build_afl and build_libfuzzer should produce libFuzzingEngine.a or some such, then build.sh should link against it.

File: common.sh [hunk 2]
```diff
@@ -32,6 +48,23 @@ get_svn_revision() {
...
+  ar r $LIB_FUZZING_ENGINE *.o
+  rm *.o
+
+  BINARY_NAME_EXT="_${FUZZER}"
+}
+
 build_libfuzzer() {
   $LIBFUZZER_SRC/build.sh
+  #mv libFuzzer.a $LIB_FUZZING_ENGINE # more consistent style, breaks backwards compatibility
+  LIB_FUZZING_ENGINE="libFuzzer.a"
```
Comment (kcc) [inline common.sh]: maybe don't do this, instead change all the build.sh files to use $LIB_FUZZING_ENGINE

File: lcms-2017-03-21/build.sh [hunk 3]
```diff
@@ -1,16 +1,17 @@
 #!/bin/bash
-# Copyright 2016 Google Inc. All Rights Reserved.
+# Copyright 2017 Google Inc. All Rights Reserved.
 # Licensed under the Apache License, Version 2.0 (the "License");
-. $(dirname $0)/../common.sh
+
+. $(dirname $0)/../common.sh $1
```
Comment (kcc) [inline lcms-2017-03-21/build.sh]: remove $1
instead, use env var, e.g. FUZZING_ENGINE
Default (empty) means libFuzzer

**Comment by kcc** [comment]
why did you chmod +x common.sh?

**Comment by kcc** [comment]
and this seems to have broken some becnhamrks, at least re2

**Comment by kcc** [comment]
Hopefully fixed in 30014fbe1b892f9b61a93dd5ec06f494ffd152e2

**Comment by morehouse** [comment]
Since merge, OpenSSL 1.0.1f is also failing to build for me.

**Comment by kcc** [comment]
@morehouse do you have details? 
The bot is green. http://lab.llvm.org:8011/builders/sanitizer-x86_64-linux-fuzzer/builds/6521
Note that this version of openssl has a bug in Makefiles that causes the parallel build to be a bit flaky

**Comment by morehouse** [comment]
@kcc That's probably what I ran into.  Just rebuilt and it worked.  Should have looked at the README first.

**Pull request merged**